src/cache/algorithms/expr_variable_reordering.cpp

```diff
@@ -0,0 +1,253 @@
+  expr2t::expr_ids::lessthan_id,
+  expr2t::expr_ids::greaterthan_id,
+  expr2t::expr_ids::lessthanequal_id,
+  expr2t::expr_ids::greaterthanequal_id};
+} // namespace
+
+void expr_variable_reordering::run()
+{
+  auto expr_type = get_expr_type(expr);
+  switch(expr_type)
+  {
+  case PARSE_AS::BIN_OP:
+  {
+    this->run_on_binop(expr);
+    break;
+  }
+  case PARSE_AS::RELATION:
+  {
+    this->run_on_relation(expr);
+    break;
+  }
+  default:;
+  }
+}
+expr_variable_reordering::PARSE_AS
+expr_variable_reordering::get_expr_type(expr2tc &expr)
+{
+  static std::map<expr2t::expr_ids, PARSE_AS> expr_to_function;
+  static bool map_initialized = false;
+
+  if(!map_initialized)
+  {
+    for(auto i : binary_operations)
+    {
+      expr_to_function[i] = PARSE_AS::BIN_OP;
+    }
+
+    for(auto i : relations)
+    {
+      expr_to_function[i] = PARSE_AS::RELATION;
+    }
+
+    expr_to_function[expr2t::expr_ids::constant_int_id] = PARSE_AS::CONSTANT;
+
+    expr_to_function[expr2t::expr_ids::symbol_id] = PARSE_AS::SYMBOL;
+    map_initialized = true;
+  }
+
+  if(expr_to_function.find(expr->expr_id) == expr_to_function.end())
+  {
+    return PARSE_AS::SKIP;
+  }
+  return expr_to_function[expr->expr_id];
+}
+
+void expr_variable_reordering::run_on_binop(expr2tc &expr) noexcept
+{
+  std::shared_ptr<arith_2ops> arith;
+  arith = std::dynamic_pointer_cast<arith_2ops>(expr);
+  // If arith is false then the cast failed
+  assert(arith);
+
+  /** Reorder both sides recursively
+   *
+   * The idea of this algorithm is to parse all symbols and values of a
+   * expression with the same operator and rewrite it.
+   *
+   * 1. Create a list of symbols and values
+   * 2. Order it in crescent order
+   * 3. Change all values
+   */
+
+  expr_variable_reordering side1_reordering(arith->side_1);
+  expr_variable_reordering side2_reordering(arith->side_2);
+
+  side1_reordering.run();
+  side2_reordering.run();
+
+  symbols_vec symbols; // annotates all inner symbols
+  values_vec values;   // annotates all inner values
+
+  // 1. Create a list of symbols and values
+  this->transverse_read_binop(arith, symbols, values);
+
+  // 2. Order it in crescent order
+  sort(symbols.begin(), symbols.end(), [](const auto &lhs, const auto &rhs) {
+    return lhs->get_symbol_name() > rhs->get_symbol_name();
+  });
+
+  // A constant propagation should have been executed prior to this
+  assert(values.size() <= 1);
+
+  // 3. Change all values
+  this->transverse_replace_binop(arith, symbols, values);
+}
+
+void expr_variable_reordering::run_on_relation(expr2tc &expr) noexcept
+{
+  std::shared_ptr<relation_data> relation;
+  relation = std::dynamic_pointer_cast<relation_data>(expr);
+
+  // If relation is false then the cast failed
+  assert(relation);
+
+  expr_variable_reordering side1(relation->side_1);
+  side1.run();
+
+  expr_variable_reordering side2(relation->side_2);
+  side2.run();
+}
+
+inline void expr_variable_reordering::transverse_read_binop(
```

Comment:
Add short description for all methods.

Reply:
The header includes the description of the method